maestrowf/interfaces/script/localparscriptadapter.py

```diff
@@ -0,0 +1,461 @@
+###############################################################################
+# Copyright (c) 2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory
+# Written by Francesco Di Natale, dinatale3@llnl.gov.
+#
+# LLNL-CODE-734340
+# All rights reserved.
+# This file is part of MaestroWF, Version: 1.0.0.
+#
+# For details, see https://github.com/LLNL/maestrowf.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+###############################################################################
+
+"""Local interface implementation."""
+import logging
+import os
+import psutil
+import signal
+import uuid
+
+from concurrent.futures import ThreadPoolExecutor
+from concurrent.futures._base import PENDING as future_PENDING
+from functools import partial as func_partial
+from threading import Thread, RLock
+
+from maestrowf.abstracts.enums import JobStatusCode, SubmissionCode, \
+    CancelCode, State
+from maestrowf.interfaces.script import CancellationRecord, SubmissionRecord
+from maestrowf.abstracts.interfaces import SchedulerScriptAdapter
+from maestrowf.abstracts import Singleton
+from maestrowf.utils import start_process
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class LocalParallelScriptAdapter(SchedulerScriptAdapter):
```

Comment:
I'm looking over this and it could be because I haven't looked at it/talked with you about it in a while, but it looks like the adapter in its current form doesn't do any substitutions. It's entirely reliant on hard-scripting of `srun` and other MPI calls in the steps themselves.

I think we talked about this a while back -- but this is where we could write a new adapter that takes in as an argument a scheduler type and then "borrows" its parallelization methods. In that same respect, we could bridge a little further and allow that to dictate scheduling Maestro the cluster essentially. That would come down to a hard check somewhere, so I'd have to mock something up. For reference, here's a mock local scheduler-esque type class I wrote in pyaestro: [`Executor` class](https://github.com/FrankD412/pyaestro/blob/master/pyaestro/utilities/executor.py#L41)

Reply:
Right, this is essentially just the replacement for the local adapter.  So it's more suited to being the base class for the other scheduler adapters.  So one extra thing that this setup might want for running on clusters is some extension of the input language to allow specifying multiple allocations to run in parallel and splitting jobs across them in some fashion.  Unsure if that's a good thing to add in the initial pr though as that could be a lot of work to make it robust/user friendly.

Reply:
On second thought, that allocation splitting thing is definitely not a good plan in this pr.

Reply:
> On second thought, that allocation splitting thing is definitely not a good plan in this pr.

I'm not sure I remember the specifics of this one. Can you refresh me?

Reply:
That's more of a next stage or even next layer up in the ecosystem: the idea of orchestrating one workflow across multiple allocations at once, so a bit beyond the scope of the threaded adapters

Reply:
Ah -- got it.

Reply:
So, digging around in these adapters again, there seems to be a few paths to consider for launcher token enabled version of this adapter -> make a parallelschedulerscript adapter and fork off variants of the concrete instances from that, or make that schedulerscriptadapter the base for everything to bring in the launcher token replacement machinery.  the latter case would potentially allow the one adapter to accept a scheduler type key so it can go use the get_parallelize_command from some dummy instances of the concrete scheduler adapters.  so, it would seem the design question is have one potentially more complicated adapter or double the total number of adapter types to enable scheduler specific tweaks in this parallel version -> so slurmparalleladapter, fluxparalleladapter (maybe this one's not needed?), etc..